python/ray/dag/compiled_dag_node.py

```diff
@@ -386,6 +379,116 @@ def __init__(
             assert not isinstance(val, ChannelInterface)
             assert not isinstance(val, DAGInputAdapter)
 
+        # Input reader to read input data from upstream DAG nodes.
+        self.input_reader: ReaderInterface = SynchronousReader(self.input_channels)
+        # Output writer to write output data to downstream DAG nodes.
+        self.output_writer: WriterInterface = SynchronousWriter(self.output_channel)
+        # Store the intermediate result of a READ or COMPUTE operation.
+        # The result of a READ operation will be used by a COMPUTE operation,
+        # and the result of a COMPUTE operation will be used by a WRITE operation.
+        self._cache = None
```

Comment:
Add types?

Reply:
also cache sounds like a bad name. maybe call it intermediate_buffer?

Reply:
Updated https://github.com/ray-project/ray/pull/46911/commits/8667ed48d4c5724e9da69035864e05c501c61d56